packages/core/src/browser/preferences/preference-contribution.ts

```diff
@@ -102,22 +135,36 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
         return this.validateFunction({ [name]: value }) as boolean;
     }
 
-    getCombinedSchema(): PreferenceSchema {
+    getCombinedSchema(): PreferenceDataSchema {
         return this.combinedSchema;
     }
 
-    getPreferences(): { [name: string]: any } {
-        return this.preferences;
-    }
-
     setSchema(schema: PreferenceSchema): void {
         this.doSetSchema(schema);
         this.updateValidate();
-        this.fireOnDidPreferencesChanged();
+        for (const property of Object.keys(schema.properties)) {
+            const schemaProps = schema.properties[property];
+            this.emitPreferenceChangedEvent(property, schemaProps.default, undefined);
+        }
```

Comment:
@elaihau here new changes to `PreferencesFrontendApplicationContribution`: https://github.com/theia-ide/theia/pull/4003/files#diff-9edbd6a2baf11d0ba53136cf40cb9ab7R40

Reply:
`this.emitPreferenceChangedEvent()` in this change calls `this.onDidPreferencesChangedEmitter.fire()`, so it wouldn't break PR 4003.

Reply:
It breaks because before reconciliation was happening only once for all properties not for each. It is expensive now. You set a schema with 3 new preferences and instead of serializing the combined scheme once, it would be serialized 3 times.

Reply:
Got it. I reverted my changes to `PreferenceServiceImpl.reconcilePreferences()`, and bring back `onPreferencesChanged`.

In the current version, I have both `onPreferenceChanged` for every single settings change, and `onPreferencesChanged` for clients that need bulk updates.